### ISSUE TITLE
Misleading error message when incorrectly assigning to a declared const/let

### ISSUE THREAD

**Issue by DanKaplanSES** [microsoft/TypeScript#56037]
### 🔎 Search Terms

Error Message, declare keyword

https://github.com/microsoft/TypeScript/issues/26118 may be relevant.

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about common bugs that aren't bugs.

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.0-dev.20231008#code/CYUwxgNghgTiAEYD2A7AzgF3iAHgRngF54AKAOgoAdYoBbNALngFcUBrFJAdxQG0BdAJREAfPABuSAJbAA3AChQkWAgggsuAExFSFMtRh1GLdpx4DhhMZJmygA

### 💻 Code

```ts
declare const ex1 = (...params: unknown[]) => void;
declare let ex2 = (...params: unknown[]) => void;
```

### 🙁 Actual behavior

An error message that says, "Expression expected. (1109)". The error is on the last character (the `;`).

![image](https://github.com/microsoft/TypeScript/assets/373675/9d57a79e-b83a-453c-b780-741a0a2e409b)

### 🙂 Expected behavior

An error message that says something like, "Expected a type but received a value instead." The error should be on the `=` character.

### Additional information about the issue

I'm not 100% sure this is a bug because "expression" may mean something different in the context of typescript's type system. But the error message seems like a bug when looking at [the JavaScript definition of "expression."](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_operators)

**Comment by jakebailey** [comment]
Did you mean:


```ts
declare const ex1: (...params: unknown[]) => void;
declare let ex2: (...params: unknown[]) => void;
```

`declare` is TypeScript specific syntax for declaring that a variable exists, but those declarations don't actually emit any code.

**Comment by jakebailey** [comment]
Oh, you mean that there should be a better error for this?

**Comment by RyanCavanaugh** [comment]
> I'm not 100% sure this is a bug because "expression" may mean something different in the context of typescript's type system. But the error message seems like a bug when looking at

Initializers, even in ambient contexts, follow the general expression grammar, so you're sort of in the progress of writing a valid program. Had you written

```
declare let ex2 = (...params: unknown[]) => void 2;
```
then you'd have a grammatically-legal program with a semantic error that you can't use that kind of initializer. In this context, `2` *is* an expression, and is parsed as such.

**Comment by DanKaplanSES** [comment]
@RyanCavanaugh Fair enough (TBH, I don't know how to parse that example but that's beside the point). Would it make sense to convert this issue from a bug to a feature request? A feature request to improve the error message when writing my original examples?

**Comment by fatcerberus** [comment]
> TBH, I don't know how to parse that example

It’s an arrow function that takes any number of `unknown` arguments and returns `undefined` (i.e. `void 2`).  Basically `declare let foo = valueExpr` is considered grammatically legal and only produces an error later in compilation.  So in *your* example you get “expected expression” because it’s trying to parse the function type as an arrow function and is expecting an expression to come after `void`.  You get exactly the same error, for the same reasons, if you delete the `declare`.

The error message can’t really be improved without making `declare let foo = …` an early syntax error.

**Comment by RyanCavanaugh** [comment]
The problem is basically this:
```ts
const a = 1; // OK
const b = void 2; // OK
const c = void; // Missing void operand, error is correct
declare const c = 1; // OK
declare const d = void 2; // Syntax OK but not a legal initializer
declare const e = () => void; // Missing void operand, error is misleading
declare const f = () => void 3; // Also not a legal initializer
```
At the point that `e` realizes there's a missing `void` operand, it's realistically "too late" to unwind the stack all the way back to the "Let's parse an initializer expression"; it's the same state that `c` is in (several stack frames deep).

Unless this kind of error were exceptionally common, it's not worth adding a fairly large amount of logic to the parser to detect it

**Comment by fatcerberus** [comment]
I guess the implied improvement request is to  just *not* do the

> Let's parse an initializer expression

step at all for `declare` vars and syntax-error at the `=`, since it’s never legal for them to have initializers anyway.
